File: webside/src/utils/mgmtIdCipher.js
/**
 * 管理番号暗号（低位在前），两种编码：
 * - 五进制（默认）：-=~<> 对应 0–4
 * - 二进制：菱形 ◇◆ 对应 0–1（◇=0，◆=1）
 *
 * 当前编码模式记录在后端 config 表（隐藏页 /x9 切换），启动时拉取并 setCipherMode()。
 * 编码与解析都**严格按当前模式**：选了五进制就只按五进制解析（非兼容解析）。
 * 与 backend/src/use_mercari/mgmt_id_cipher.py 保持一致。
 */

export const MGMT_CIPHER_ALPHABET = '-=~<>' // 五进制 0–4
export const MGMT_BINARY_ALPHABET = '◇◆' // 二进制 0–1（◇=0，◆=1）

const BASE5_SET = new Set(MGMT_CIPHER_ALPHABET.split(''))
const BINARY_SET = new Set(MGMT_BINARY_ALPHABET.split(''))
const BASE5_TOKEN_RE = /^([-=~<>]+)(?:\s*[*xX×]\s*(\d+))?$/u
const BINARY_TOKEN_RE = /^([◇◆]+)(?:\s*[*xX×]\s*(\d+))?$/u

// 当前编码模式：'binary' | 'base5'（默认 base5）。由 setCipherMode 在启动/切换时更新。
let currentMode = 'base5'

export function setCipherMode(mode) {
  currentMode = String(mode || '').toLowerCase() === 'binary' ? 'binary' : 'base5'
  return currentMode
}

export function getCipherMode() {
  return currentMode
}

function modeSpec(mode) {
  const m = (mode || currentMode) === 'binary' ? 'binary' : 'base5'
  return m === 'binary'
    ? { alphabet: MGMT_BINARY_ALPHABET, base: 2, charset: BINARY_SET, tokenRe: BINARY_TOKEN_RE }
    : { alphabet: MGMT_CIPHER_ALPHABET, base: 5, charset: BASE5_SET, tokenRe: BASE5_TOKEN_RE }
}

export function encodeMgmtId(value, mode = currentMode) {
  let n = Number(value)
  if (!Number.isFinite(n) || n < 0) {
    throw new Error('invalid management id')
  }
  n = Math.floor(n)
  const { alphabet, base } = modeSpec(mode)
  if (n === 0) return alphabet[0]
  const chars = []
  while (n > 0) {
    chars.push(alphabet[n % base])
    n = Math.floor(n / base)
  }
  return chars.join('')
}

export function decodeMgmtIdCipher(token, mode = currentMode) {
  const s = String(token || '').trim()
  if (!s) return null
  const { alphabet, base, charset } = modeSpec(mode)
  if ([...s].some((c) => !charset.has(c))) return null
  let n = 0
  let mult = 1
  for (const c of s) {
    n += alphabet.indexOf(c) * mult
    mult *= base
  }
  return n > 0 ? n : null
}

export function encodeMgmtIds(ids, sep = '、') {
  const mode = currentMode
  const parts = []
  for (const raw of ids || []) {
    const n = Number(raw)
    if (Number.isFinite(n) && n > 0) parts.push(encodeMgmtId(n, mode))
  }
  return parts.join(sep)
}

function splitCipherChunks(segment) {
  return String(segment || '')
    .split(/[,，、\s]+/)
    .map((p) => p.trim())
    .filter(Boolean)
}

function cipherTokenBaseAndQty(token, mode) {
  const t = String(token || '').trim()
  if (!t) return { base: '', qty: 1 }
  const { tokenRe } = modeSpec(mode)
  const m = t.match(tokenRe)
  if (!m) return { base: '', qty: 1 }
  const base = (m[1] || '').trim()
  const qraw = (m[2] || '').trim()
  if (!qraw) return { base, qty: 1 }
  const q = parseInt(qraw, 10)
  return { base, qty: Number.isFinite(q) && q > 0 ? q : 1 }
}

export function isCipherMgmtLine(line, mode = currentMode) {
  const s = String(line || '').trim()
  if (!s) return false
  if (/管理\s*(?:ID|番号)\s*[:：]/iu.test(s)) return false
  if (/バーコード\s*[:：]/u.test(s)) return false
  const { charset } = modeSpec(mode)
  let hasToken = false
  for (const part of splitCipherChunks(s)) {
    const { base } = cipherTokenBaseAndQty(part, mode)
    if (!base || [...base].some((c) => !charset.has(c))) return false
    hasToken = true
  }
  return hasToken
}

/** 从说明最末非空行解析暗号管理番号（严格按当前模式），返回 [{ id, quantity }] */
export function parseTrailingCipherMgmtIds(text) {
  const s = String(text ?? '').trim()
  if (!s) return []
  const mode = currentMode
  const lines = s.split(/\r?\n/)
  let lastLine = ''
  for (let i = lines.length - 1; i >= 0; i--) {
    const t = String(lines[i] || '').trim()
    if (t) {
      lastLine = t
      break
    }
  }
  if (!lastLine || !isCipherMgmtLine(lastLine, mode)) return []
  const out = []
  for (const part of splitCipherChunks(lastLine)) {
    const { base, qty } = cipherTokenBaseAndQty(part, mode)
    if (!base) continue
    const id = decodeMgmtIdCipher(base, mode)
    if (id != null) out.push({ id, quantity: qty })
  }
  return out
}

/** 合并去重后的管理番号 id 列表（仅 id，不含数量） */
export function parseMgmtIdsFromDescription(text) {
  const ids = []
  const seen = new Set()
  for (const { id } of parseTrailingCipherMgmtIds(text)) {
    if (!seen.has(id)) {
      seen.add(id)
      ids.push(id)
    }
  }
  return ids
}

export function stripTrailingMgmtBlock(text) {
  let s = String(text || '')
  let prev = null
  while (prev !== s) {
    prev = s
    s = s.replace(/(?:\n{1,2})?管理番号[:：][\d、,，\s]+$/u, '').trimEnd()
  }
  const lines = s.split(/\r?\n/)
  while (lines.length) {
    const last = String(lines[lines.length - 1] || '').trim()
    if (!last) {
      lines.pop()
      continue
    }
    if (isCipherMgmtLine(last)) {
      lines.pop()
      continue
    }
    break
  }
  return lines.join('\n').trimEnd()
}

File: backend/src/use_mercari/mgmt_id_cipher.py
# -*- coding: utf-8 -*-
"""
管理番号暗号（低位在前），两种编码：
- 五进制（默认）：字符集 ``-=~<>`` 对应 0–4。
- 二进制：菱形 ``◇◆`` 对应 0–1（◇=0，◆=1）。

当前编码模式记录在 config 表键 ``mgmt_cipher_mode``（'binary' / 'base5'，默认 base5），
由隐藏页 /x9 切换。**编码与解析都严格按当前模式**：选了五进制就只按五进制解析，
反之亦然（非兼容解析）。出品时写在商品说明最末行，无「管理番号：」前缀。
"""

from __future__ import annotations

import re
from typing import List, Optional, Tuple

MGMT_CIPHER_ALPHABET = "-=~<>"   # 五进制 0–4
MGMT_BINARY_ALPHABET = "◇◆"      # 二进制 0–1（◇=0，◆=1）
MGMT_CIPHER_MODE_KEY = "mgmt_cipher_mode"

_BASE5_SET = set(MGMT_CIPHER_ALPHABET)
_BINARY_SET = set(MGMT_BINARY_ALPHABET)
# 各模式各自的 token 正则（'-' 置于字符类开头表示字面量）
_BASE5_TOKEN_RE = re.compile(r"^([-=~<>]+)(?:\s*[*xX×]\s*(\d+))?$", re.UNICODE)
_BINARY_TOKEN_RE = re.compile(r"^([◇◆]+)(?:\s*[*xX×]\s*(\d+))?$", re.UNICODE)


def get_cipher_mode() -> str:
    """读取当前编码模式：'binary' 或 'base5'（默认 base5）。"""
    try:
        from ..db_manage.models.config_entry import ConfigEntryModel

        v = (ConfigEntryModel.get_value(MGMT_CIPHER_MODE_KEY) or "").strip().lower()
    except Exception:  # noqa: BLE001
        v = ""
    return "binary" if v == "binary" else "base5"


def _mode_spec(mode: Optional[str] = None):
    """返回 (alphabet, base, charset, token_re)；mode 省略时取当前配置模式。"""
    m = mode or get_cipher_mode()
    if m == "binary":
        return MGMT_BINARY_ALPHABET, 2, _BINARY_SET, _BINARY_TOKEN_RE
    return MGMT_CIPHER_ALPHABET, 5, _BASE5_SET, _BASE5_TOKEN_RE


def encode_mgmt_id(value: int, mode: Optional[str] = None) -> str:
    """将正整数 inventory.id 编码为暗号串（mode 省略时取当前配置模式）。"""
    try:
        n = int(value)
    except (TypeError, ValueError):
        raise ValueError("invalid management id") from None
    if n < 0:
        raise ValueError("invalid management id")
    alphabet, base, _set, _re = _mode_spec(mode)
    if n == 0:
        return alphabet[0]
    chars: List[str] = []
    while n > 0:
        chars.append(alphabet[n % base])
        n //= base
    return "".join(chars)


def decode_mgmt_id_cipher(token: str, mode: Optional[str] = None) -> Optional[int]:
    """解码单个暗号 token；严格按当前(或指定)模式的字符集解析，非法返回 None。"""
    s = (token or "").strip()
    if not s:
        return None
    alphabet, base, charset, _re = _mode_spec(mode)
    if not all(c in charset for c in s):
        return None
    n = 0
    mult = 1
    for c in s:
        n += alphabet.index(c) * mult
        mult *= base
    return n if n > 0 else None


def encode_mgmt_ids(ids: List[int], sep: str = "、") -> str:
    mode = get_cipher_mode()  # 一次读取，避免逐个查库
    parts: List[str] = []
    for raw in ids:
        try:
            n = int(raw)
        except (TypeError, ValueError):
            continue
        if n > 0:
            parts.append(encode_mgmt_id(n, mode=mode))
    return sep.join(parts)


def _split_cipher_chunks(segment: str) -> List[str]:
    parts: List[str] = []
    for part in re.split(r"[,，、\s]+", segment or ""):
        p = (part or "").strip()
        if p:
            parts.append(p)
    return parts


def _cipher_token_base_and_qty(token: str, mode: Optional[str] = None) -> Tuple[str, int]:
    t = (token or "").strip()
    if not t:
        return "", 1
    _a, _b, _set, token_re = _mode_spec(mode)
    m = token_re.match(t)
    if not m:
        return "", 1
    base = (m.group(1) or "").strip()
    qraw = (m.group(2) or "").strip()
    if not qraw:
        return base, 1
    try:
        q = int(qraw)
    except (TypeError, ValueError):
        q = 1
    return base, max(1, q)


def is_cipher_mgmt_line(line: str, mode: Optional[str] = None) -> bool:
    """整行是否仅由「当前模式」的暗号 token（及分隔符）组成。"""
    s = (line or "").strip()
    if not s:
        return False
    if re.search(r"管理\s*(?:ID|番号)\s*[:：]", s, re.IGNORECASE):
        return False
    if re.search(r"バーコード\s*[:：]", s, re.IGNORECASE):
        return False
    m = mode or get_cipher_mode()
    _a, _b, charset, _re = _mode_spec(m)
    has_token = False
    for part in _split_cipher_chunks(s):
        base, _qty = _cipher_token_base_and_qty(part, mode=m)
        if not base or not all(c in charset for c in base):
            return False
        has_token = True
    return has_token


def parse_trailing_cipher_mgmt_tokens(text: Optional[str]) -> List[Tuple[int, int]]:
    """
    从说明**最末非空行**解析暗号管理番号（严格按当前模式）。
    返回 [(inventory_id, quantity), ...]
    """
    if text is None:
        return []
    s = str(text).strip()
    if not s:
        return []

    mode = get_cipher_mode()  # 一次读取，整次解析保持一致
    lines = s.splitlines()
    last_line = ""
    for raw in reversed(lines):
        t = str(raw or "").strip()
        if t:
            last_line = t
            break
    if not last_line or not is_cipher_mgmt_line(last_line, mode=mode):
        return []

    out: List[Tuple[int, int]] = []
    for part in _split_cipher_chunks(last_line):
        base, qty = _cipher_token_base_and_qty(part, mode=mode)
        if not base:
            continue
        mid = decode_mgmt_id_cipher(base, mode=mode)
        if mid is None:
            continue
        out.append((mid, qty))
    return out
